src/renderers/dom/shared/__tests__/ReactDOMServerIntegration-test.js

```diff
@@ -410,4 +437,578 @@ describe('ReactDOMServerIntegration', () => {
       expect(e.getAttribute('click')).toBe(null);
     });
   });
+
+  describe('elements and children', function() {
+    // helper functions.
+    const TEXT_NODE_TYPE = 3;
+    const COMMENT_NODE_TYPE = 8;
+
+    function expectNode(node, type, value) {
+      expect(node).not.toBe(null);
+      expect(node.nodeType).toBe(type);
+      expect(node.nodeValue).toMatch(value);
+    }
+
+    function expectTextNode(node, text) {
+      if (ReactDOMFeatureFlags.useFiber) {
+        // with Fiber, a React text node is just a DOM text node.
+        expectNode(node, TEXT_NODE_TYPE, text);
+      } else {
+        // with Stack, a React text node is a DOM text node surrounded by
+        // react-text comment nodes.
+        expectNode(node, COMMENT_NODE_TYPE, / react-text: [0-9]+ /);
+        if (text.length > 0) {
+          node = node.nextSibling;
+          expectNode(node, TEXT_NODE_TYPE, text);
+        }
+        expectNode(node.nextSibling, COMMENT_NODE_TYPE, / \/react-text /);
+      }
+    }
+
+    function expectEmptyNode(node) {
+      expectNode(node, COMMENT_NODE_TYPE, / react-empty: [0-9]+ /);
+    }
+
+    describe('text children', function() {
+      itRenders('a div with text', async render => {
+        const e = await render(<div>Text</div>);
+        expect(e.tagName).toBe('DIV');
+        expect(e.childNodes.length).toBe(1);
+        expectNode(e.firstChild, TEXT_NODE_TYPE, 'Text');
+      });
+
+      itRenders('a div with text with flanking whitespace', async render => {
+        // prettier-ignore
+        const e = await render(<div>  Text </div>);
+        expect(e.childNodes.length).toBe(1);
+        expectNode(e.childNodes[0], TEXT_NODE_TYPE, '  Text ');
```

Comment:
I think we only collapse leading and trailing whitespace, but only if it's leading or trailing for the whole line.